jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java

```diff
@@ -0,0 +1,217 @@
+        queue = (MonitoringBlockingArrayQueue)getQueue();
+        setStopTimeout(2000);
+    }
+
+    @Override
+    public void execute(final Runnable job)
+    {
+        final long begin = System.nanoTime();
+        super.execute(new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                long queueLatency = System.nanoTime() - begin;
+                tasks.increment();
+                Atomics.updateMax(maxQueueLatency, queueLatency);
+                totalQueueLatency.add(queueLatency);
+                Atomics.updateMax(maxThreads, threads.incrementAndGet());
+                long start = System.nanoTime();
+                try
+                {
+                    job.run();
+                }
+                finally
+                {
+                    long taskLatency = System.nanoTime() - start;
+                    threads.decrementAndGet();
+                    Atomics.updateMax(maxTaskLatency, taskLatency);
+                    totalTaskLatency.add(taskLatency);
+                }
+            }
+
+            @Override
+            public String toString()
+            {
+                return job.toString();
+            }
+        });
+    }
+
+    @ManagedOperation(value = "resets the statistics", impact = "ACTION")
+    public void reset()
+    {
+        tasks.reset();
+        maxTaskLatency.set(0);
+        totalTaskLatency.reset();
+        queue.reset();
+        maxQueueLatency.set(0);
+        totalQueueLatency.reset();
+        threads.set(0);
+        maxThreads.set(0);
+    }
+
+    @ManagedAttribute("the number of tasks executed")
+    public long getTasks()
+    {
+        return tasks.sum();
+    }
+
+    @ManagedAttribute("the maximum number of busy threads")
+    public int getMaxBusyThreads()
+    {
+        return maxThreads.get();
+    }
+
+    @ManagedAttribute("the maximum task queue size")
+    public int getMaxQueueSize()
+    {
+        return queue.maxSize.get();
+    }
+
+    @ManagedAttribute("the average time a task remains in the queue, in nanoseconds")
+    public long getAverageQueueLatency()
+    {
+        long count = tasks.sum();
+        return count == 0 ? -1 : totalQueueLatency.sum() / count;
+    }
+
+    @ManagedAttribute("the maximum time a task remains in the queue, in nanoseconds")
+    public long getMaxQueueLatency()
+    {
+        return maxQueueLatency.get();
+    }
+
+    @ManagedAttribute("the average task execution time, in nanoseconds")
+    public long getAverageTaskLatency()
+    {
+        long count = tasks.sum();
+        return count == 0 ? -1 : totalTaskLatency.sum() / count;
+    }
+
+    @ManagedAttribute("the maximum task execution time, in nanoseconds")
+    public long getMaxTaskLatency()
+    {
+        return maxTaskLatency.get();
+    }
+
+    public static class MonitoringBlockingArrayQueue extends BlockingArrayQueue<Runnable>
```

Comment:
why is this needed?  If you use a CounterStatistics class for the tasks as they are queued and then run, you will already know the current and max size of the queue without double counting.